From 6b0a973e52eb5d55cfa8896d528664a0b88988b9 Mon Sep 17 00:00:00 2001
From: Xiang Zhou <x245zhou@uwaterloo.ca>
Date: Tue, 15 Jun 2021 18:22:11 -0400
Subject: [PATCH] patch(bug-eg):revert to buggy version of bb_is_valid

---
 source/byte_buf.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/source/byte_buf.c b/source/byte_buf.c
index 4c84c72..afd21dc 100644
--- a/source/byte_buf.c
+++ b/source/byte_buf.c
@@ -58,7 +58,7 @@ int aws_byte_buf_init_copy(struct aws_byte_buf *dest, struct aws_allocator *allo
 bool aws_byte_buf_is_valid(const struct aws_byte_buf *const buf) {
     return buf != NULL &&
            ((buf->capacity == 0 && buf->len == 0 && buf->buffer == NULL) ||
-            (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->capacity)));
+            (buf->capacity > 0 && buf->len <= buf->capacity && AWS_MEM_IS_WRITABLE(buf->buffer, buf->len)));
 }
 
 bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
-- 
2.17.1

